buildtools/main.cc

```diff
@@ -141,15 +146,89 @@ class ErrorCollector : public google::protobuf::compiler::MultiFileErrorCollecto
     void Clear() { errors_.clear(); warnings_.clear(); }
 };
 
+// Error collector for Parser (io::ErrorCollector interface)
+class ParserErrorCollector : public google::protobuf::io::ErrorCollector {
+private:
+    std::vector<std::string> errors_;
+    std::string filename_;
+
+public:
+    explicit ParserErrorCollector(const std::string& filename) : filename_(filename) {}
+
+    void RecordError(int line, int column, absl::string_view message) override {
+        std::stringstream ss;
+        ss << filename_;
+        if (line >= 0) {
+            ss << ":" << (line + 1);  // line is 0-indexed, display as 1-indexed
+            if (column >= 0) {
+                ss << ":" << (column + 1);
+            }
+        }
+        ss << " " << message;
+        errors_.push_back(ss.str());
+    }
+
+    const std::vector<std::string>& GetErrors() const { return errors_; }
+    bool HasErrors() const { return !errors_.empty(); }
+};
+
 int main(int argc, char** argv) {
     if (argc < 2) {
-        std::cerr << "Usage: " << argv[0] << " <descriptors | grpc-java>\n";
+        std::cerr << "Usage: " << argv[0] << " <descriptors | java | grpc-java | validate-syntax>\n";
         return 1;
     }
 
     std::string option = argv[1]; // Full string
 
-    if (option == "descriptors") {
+    if (option == "validate-syntax") {
```

Comment:
As discussed and not needed in this PR, but we need to improve the code structure for C++ and, maybe, start to expose the API through Wasm functions instead of using the "main", it will be faster to execute and cleaner even if we loose stdin/stdout/stderr support